[TF-TRT] Add LogSoftmax Support for TF-TRT

pavanimajety opened the pull request:
This PR adds TF-TRT support to `tf.nn.log_softmax` operation.  This is performed using the formula `logsoftmax = logits - log(reduce_sum(exp(logits), axis=-1))` .  The implemented TRT layers are fused into a single op.   

@DEKHTIARJonathan @tfeher : Please review the changes.

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -4721,6 +4721,58 @@ Status ConvertSoftmax(OpConverterParams* params) {
+  if (params->validation_only) return Status::OK();
+
+  // Perform LogSoftmax operation -
+  // logsoftmax = logits - log(reduce_sum(exp(logits), axis))
```
Is there a reason to prefer this formulation over  `log(softmax(logits))`?

pavanimajety replied:
I chose this way because tensorflow implements in Python using `logsoftmax = logits - log(reduce_sum(exp(logits), axis))` .

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -4721,6 +4721,58 @@ Status ConvertSoftmax(OpConverterParams* params) {
   return Status::OK();
 }
 
+Status ConvertLogSoftmax(OpConverterParams* params) {
```
Instead of adding new converters to convert_nodes.cc, please prefer to:
- derive it from [OpConverterBase](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/op_converter.h)
- place it under [convert/ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/ops)
- Place the validation code into `Validate()`, and the actual conversion into `Convert()`, see the [tile converter](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc) as an example
- don't forget to add an entry to [BUILD](https://github.com/tensorflow/tensorflow/blob/cf29afd3e937e65cdc8d4dff824bcc0af09ab43d/tensorflow/compiler/tf2tensorrt/BUILD#L682)

tfeher reviewed (changes requested):
Hi Pavani, thank you for the PR, it looks great overall! But for new converters we prefer to derive them from `OpConverterBase`, I have left instructions on how to update.

tfeher reviewed (approved):
Thanks @pavanimajety for refactoring the converter, the PR looks good to me!

bixia1 commented:
@pavanimajety Please add a PR description.

bixia1 reviewed (changes requested):
Looks good to me, just request fix to comment style and a PR description.

pavanimajety commented on tensorflow/compiler/tf2tensorrt/convert/ops/log_softmax.cc:
```diff
@@ -0,0 +1,94 @@
+    // Perform LogSoftmax operation -
+    // logsoftmax = logits - log(reduce_sum(exp(logits), axis))
+
+    // get the logits tensor
```
Fixed the comments.

bixia1 reviewed (commented):
Looks good to me now. Would you please squash the commits?

pavanimajety committed:
Add LogSoftmax conversion


Fix Softmax comments


[TF-TRT] Move LogSoftmax to use OpConverterBase


Fix compiler errors


clang-format

Undo changes to convert_nodes.cc


Fix comments

pavanimajety commented:
> Looks good to me now. Would you please squash the commits?

Done. Thanks for reviewing

pavanimajety commented:
Sorry, rebased for further work by mistake.

bixia1 commented:
This is merged, wonder why it doesn't show as merged here.

pavanimajety commented:
> This is merged, wonder why it doesn't show as merged here.

Would you mind checking again? I don't see in tensorflow:master : https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/tf2tensorrt/convert/ops

mihaimaruseac commented:
It did not show as merged here because copybara sync was broken. Fixed in big squash commit 49f97f135a2e1d5d22e60d2a80ec668d53f9708a
